### ISSUE TITLE
Bug: use-isnan false positive on shadowed NaN/Number

### ISSUE THREAD

**Issue by Kuldeep2822k** [eslint/eslint#20950]
### Environment

Node version: v24.15.0
npm version: v11.12.1
Local ESLint version: v10.4.1
Global ESLint version: Not installed
Operating System: Windows

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
{
    "rules": {
        "use-isnan": ["error", {
            "enforceForIndexOf": true
        }]
    }
}
```
</details>

```js
// 1. Shadowing NaN in a local scope
let NaN = 5; 
if (x === NaN) {}

// 2. Shadowing Number in a function parameter
function f(Number) { 
    return x === Number.NaN; 
}

// 3. Shadowing NaN in a switch case
function testSwitch(foo) {
    let NaN = 5; 
    switch (foo) { 
        case NaN: break; 
    }
}

// 4. Shadowing NaN inside an Array.prototype.indexOf argument
function foo(NaN) { 
    return arr.indexOf(NaN); 
}
```

### What did you expect to happen?

I expected no errors to be reported.

In the provided cases, the NaN and Number identifiers are explicitly shadowed by local variables or function parameters. Because they no longer reference the built-in global NaN, comparing against them is perfectly valid JavaScript. The rule should use scope analysis to ensure it only flags references to the true global object.

### What actually happened?

ESLint incorrectly reported false positives for all the shadowed bindings because it matched them by name only.

```text
   3:5  error  Use the isNaN function to compare with NaN.  local/use-isnan
   7:12  error  Use the isNaN function to compare with NaN.  local/use-isnan
   14:9  error  'case NaN' can never match. Use Number.isNaN before the switch.  local/use-isnan
   20:12  error  Array prototype method 'indexOf' cannot find NaN.  local/use-isnan
✖ 4 problems (4 errors, 0 warnings)
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLy8gMS4gU2hhZG93aW5nIE5hTiBpbiBhIGxvY2FsIHNjb3BlXG5sZXQgTmFOID0gNTtcbmlmICh4ID09PSBOYU4pIHt9XG5cbi8vIDIuIFNoYWRvd2luZyBOdW1iZXIgaW4gYSBmdW5jdGlvbiBwYXJhbWV0ZXJcbmZ1bmN0aW9uIHRlc3ROdW1iZXIoTnVtYmVyKSB7XG4gICAgcmV0dXJuIHggPT09IE51bWJlci5OYU47XG59XG5cbi8vIDMuIFNoYWRvd2luZyBOYU4gaW4gYSBzd2l0Y2ggY2FzZVxuZnVuY3Rpb24gdGVzdFN3aXRjaChmb28pIHtcbiAgICBsZXQgTmFOID0gNTtcbiAgICBzd2l0Y2ggKGZvbykge1xuICAgICAgICBjYXNlIE5hTjogYnJlYWs7XG4gICAgfVxufVxuXG4vLyA0LiBTaGFkb3dpbmcgTmFOIGluc2lkZSBhbiBBcnJheS5wcm90b3R5cGUuaW5kZXhPZiBhcmd1bWVudFxuZnVuY3Rpb24gdGVzdEluZGV4T2YoTmFOKSB7XG4gICAgcmV0dXJuIGFyci5pbmRleE9mKE5hTik7XG59Iiwib3B0aW9ucyI6eyJydWxlcyI6eyJ1c2UtaXNuYW4iOlsiZXJyb3IiLHsiZW5mb3JjZUZvckluZGV4T2YiOnRydWV9XX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by Abusalah0** [comment]
I was able to reproduce this on Linux with ESLint `v10.4.1`.

Environment:

- OS: Linux `7.0.10-201.fc44.x86_64`
- Node: `v24.13.1`
- npm: `11.8.0`
- ESLint: `v10.4.1`
- Parser: default Espree

Smallest reproduction:

```js
let NaN = 5;
if (x === NaN) {}
```

Config:

```js
export default [
  {
    rules: {
      "use-isnan": ["error", {
        enforceForSwitchCase: true,
        enforceForIndexOf: true,
      }],
    },
  },
];
```

Command:

```sh
npx eslint repro.js
```

Actual output:

```text
/home/abd/Desktop/eslint-testing/repro/repro.js
  2:5  error  Use the isNaN function to compare with NaN  use-isnan

✖ 1 problem (1 error, 0 warnings)
```

Expected: no `use-isnan` error, because `NaN` is a local binding in this example rather than the global `NaN` value.

I also tested the other reported patterns separately on `v10.4.1`, and each reported one error:

```js
function f(Number) {
  return x === Number.NaN;
}
```

```text
2:10  error  Use the isNaN function to compare with NaN  use-isnan
```

```js
function testSwitch(foo) {
  let NaN = 5;
  switch (foo) {
    case NaN:
      break;
  }
}
```

```text
4:5  error  'case NaN' can never match. Use Number.isNaN before the switch  use-isnan
```

```js
function testIndexOf(NaN) {
  return arr.indexOf(NaN);
}
```

```text
2:10  error  Array prototype method 'indexOf' cannot find NaN  use-isnan
```

I also checked `eslint@latest` from npm:

```sh
npm view eslint version
```

```text
10.4.1
```

Since the latest published package currently resolves to `v10.4.1`, this still reproduces on latest npm ESLint.

I took a quick look at `lib/rules/use-isnan.js` as well. The rule appears to route all of these checks through `isNaNIdentifier()`, which currently matches:

```js
astUtils.isSpecificId(nodeToCheck, "NaN") ||
astUtils.isSpecificMemberAccess(nodeToCheck, "Number", "NaN")
```

The helpers check identifier/member names and AST shape, but I did not see scope resolution in this path. That seems consistent with the false positive: local bindings named `NaN` or `Number` are treated the same as references to the global built-ins.

a possible fix would be to only report direct `NaN` when it is a global reference, and only report `Number.NaN` when the `Number` object identifier resolves to the global `Number`.

**Comment by mdjermanovic** [comment]
Confirmed, a PR is welcome.
